src/api/Nhs.Appointments.Audit/Functions/Middleware.cs

```
@@ -0,0 +1,64 @@
+using System.Reflection;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using Microsoft.Extensions.DependencyInjection;
+using Nhs.Appointments.Core;
+using Nhs.Appointments.Core.Inspectors;
+
+namespace Nhs.Appointments.Audit.Functions;
+
+public class Middleware : IFunctionsWorkerMiddleware
+{
+    public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+    {
+        var requiresAudit = GetRequiresAudit(context);
+
+        if (requiresAudit == null)
+        {
+            await next(context);
+            return;
+        }
+        
+        //TODO make sure not awaiting model being built as part of fire and forget?
+        var audit = await BuildAuditModel(context, requiresAudit);
+
+        _ = RecordAudit(audit);
+        
+        await next(context);
+    }
+
+    private async Task<Models.Audit> BuildAuditModel(FunctionContext context, RequiresAuditAttribute requiresAudit)
+    {
+        var userContextProvider = context.InstanceServices.GetRequiredService<IUserContextProvider>();
+        var requestSiteInspectorType = requiresAudit.RequestSiteInspector;
+        var userId = userContextProvider.UserPrincipal?.Claims.GetUserEmail();
+        var actionType = context.FunctionDefinition.Name;
+        var siteId = string.Empty;
+        if (context.InstanceServices.GetService(requestSiteInspectorType) is IRequestInspector requestInspector)
+        {
+            var request = await context.GetHttpRequestDataAsync();
+
+            //TODO not expecting multiple site matches in first usages??
```

The get availability endpoint is the only one that has multiple sites, it is typically called by an API user, we might need to give some thought about what we want to do in these cases, but single or default will except if this list contains many so this will be a problem on that endpoint

---

Yes the singleOrDefault was intentional so that it would fail fast if we ever audit a function that has multiple sites. Do we get ahead of this and make the 'site' audit property a collection so that we store all? Only thought was it would then not be partitionable maybe...?